### PR TITLE
Added a minimum and maximum functions

### PR DESCRIPTION
**Pull request by zaleslaw**
Some notes to the proposed solution:

- both functions are added to ndarray, not to the root namespace (it looks like we compare the current ndarray and OTHER)
- not found min/max functions for complex numbers
- I'm not sure that this is the right place for both functions, I suppose it should be moved to Arithmetic or something like this
- should we return ndarray or we could return MultiArray?

### CODE REVIEW

**Comment by devcrocod** [comment]
Complex numbers cannot be ordered so there is no min and max. But there are minBy and maxBy, where you can determine how to compare complex numbers yourself.

**Comment by devcrocod** [comment]
I use this logic when returning, if we return a new object, then NDArray, if we change the current one, then we return its same type.

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt [hunk 1]
```diff
@@ -715,6 +716,47 @@ public inline fun <T, D : Dimension, reified R : Any> MultiArray<T, D>.map(trans
     return NDArray(data, shape = shape, dim = dim)
 }
 
+
+/**
+ * Returns the element-wise minimum of array elements for [this] and [other].
+ */
+public fun <T, D : Dimension> MultiArray<T, D>.minimum(other: MultiArray<T, D>): NDArray<T, D> {
```
Comment (devcrocod) [inline multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt:723]: you can restrict the type to Number

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt [hunk 2]
```diff
@@ -715,6 +716,47 @@ public inline fun <T, D : Dimension, reified R : Any> MultiArray<T, D>.map(trans
...
+    requireArraySizes(this.size, other.size)
```
Comment (devcrocod) [inline multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt:724]: It is not clear whether we should check the coincidence of shape?

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt [hunk 3]
```diff
@@ -715,6 +716,47 @@ public inline fun <T, D : Dimension, reified R : Any> MultiArray<T, D>.map(trans
...
+    val ret = if (this.consistent) (this as NDArray).copy() else (this as NDArray).deepCopy()
```
Comment (devcrocod) [inline multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt:725]: You can already use `deepcopy`, there is a consistency check inside.

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt [hunk 4]
```diff
@@ -715,6 +716,47 @@ public inline fun <T, D : Dimension, reified R : Any> MultiArray<T, D>.map(trans
...
+    when (dtype) {
+            DataType.DoubleDataType -> (ret as NDArray<Double, D>).commonAssignOp(other.iterator() as Iterator<Double>) { a, b -> min (a, b) }
```
Comment (devcrocod) [inline multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt:727]: What formatter are you using? There is usually no space after function calls: `min(a, b)`

**Review by devcrocod — CHANGES_REQUESTED** [review]
Very good PR. Thanks!

I think it's worth discussing behavior with complex numbers.

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt (hunk 4 above)
Comment (zaleslaw) [reply multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt:727]: probably, I disabled or forgot formatter, but there is only kotlin-official-codestyle is used by myself for two last years

**Comment by zaleslaw** [comment]
@devcrocod please check the updates

**Review by devcrocod — APPROVED** [review]

**Pull request merged**